src/adapter.js

```diff
@@ -284,24 +301,60 @@ export function parse({minim, source, generateSourceMap}, done) {
     try {
       ast = yamlAst.compose(source);
     } catch (err) {
-      makeAnnotation(Annotation, SourceMap, null, parseResult,
+      makeAnnotation(Annotation, Link, SourceMap, null, parseResult,
         ANNOTATIONS.AST_UNAVAILABLE, null,
         'Input AST could not be composed, so source maps will not be available');
     }
   } else {
-    makeAnnotation(Annotation, SourceMap, null, parseResult,
+    makeAnnotation(Annotation, Link, SourceMap, null, parseResult,
       ANNOTATIONS.AST_UNAVAILABLE, null,
       'Source maps are only available with string input');
   }
 
-  $RefParser.dereference(loaded, (err, swagger) => {
+  // Some sane defaults since these are sometimes left out completely
+  if (loaded.info === undefined) {
+    loaded.info = {};
+  }
+
+  if (loaded.paths === undefined) {
+    loaded.paths = {};
+  }
+
+  // Parse and validate the Swagger document!
+  parser.validate(loaded, (err) => {
+    const swagger = parser.api;
+
     if (err) {
-      return done(err, parseResult);
+      if (swagger === undefined) {
+        return done(err, parseResult);
+      }
+
+      // Non-fatal errors, so let us try and create annotations for them and
+      // continue with the parsing as best we can.
+      if (err.details) {
+        const queue = [err.details];
+        while (queue.length) {
```

Comment:
I went with a `while` because the length may change as we `push` / `shift` / `unshift` from the queue. It could be a `for` loop as long as the updated length is checked each iteration. I guess I don't have a strong opinion here, just seemed like the right thing to use.